src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -446,17 +446,22 @@ case object SSTORE extends OpCode(0x55, 2, 0, G_zero) {
 case object JUMP extends OpCode(0x56, 1, 0, G_mid) with ConstGas {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
     val (pos, stack1) = state.stack.pop
-    //TODO: JUMPDEST validation
-    state.withStack(stack1).goto(pos.intValue)
+    if(state.program.validJumpDestinations.contains(pos.intValue))
+      state.withStack(stack1).goto(pos.intValue)
+    else
+      state.withError(InvalidJump)
   }
 }
 
 case object JUMPI extends OpCode(0x57, 2, 0, G_high) with ConstGas {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
     val (Seq(pos, cond), stack1) = state.stack.pop(2)
     val nextPos = if (!cond.isZero) pos.intValue else state.pc + 1
-    //TODO: JUMPDEST validation
-    state.withStack(stack1).goto(nextPos)
+    val isValidJump = !cond.isZero && state.program.validJumpDestinations.contains(nextPos)
+    if(isValidJump || cond.isZero)
+      state.withStack(stack1).goto(nextPos)
+    else
+      state.withError(InvalidJump)
```

Comment:
I found this quite confusing and for a long time I couldn't figure out whether it's correct. The reason is the repeated `cond.isZero` as a factor of `isValidJump` and then in the `if`.

Now before I offer a solution there's an important question to be asked here. If we don't jump, should the next instruction be JUMPDEST? In other words, should JUMPDEST always follow JUMPI?

Using evm-runner I saw this in one of my test contracts (it's the code that determines what function is to be called):

```
evm-runner> Caller.code
  0000:    PUSH1 60
  0002:    PUSH1 40
  0004:    MSTORE
  0005:    PUSH4 ff ff ff ff
  000a:    PUSH1 e0
  000c:    PUSH1 02
  000e:    EXP
  000f:    PUSH1 00
  0011:    CALLDATALOAD
  0012:    DIV
  0013:    AND
  0014:    PUSH4 38 1b aa bc
  0019:    DUP2
  001a:    EQ
  001b:    PUSH2 00 2c
  001e:    JUMPI
  001f:    DUP1
  0020:    PUSH4 f5 53 32 ab
  0025:    EQ
  0026:    PUSH2 00 4e
  0029:    JUMPI
  002a:    JUMPDEST
  002b:    INVALID
  002c:    JUMPDEST
```

As you can see after the first JUMPI (offset 001f) there is no JUMPDEST. And that code is fine because we haven't matched a function signature and we're looking at another one. Ergo, `isValidJump` is wrong.

I would suggest the following rewrite:

```scala
    val (Seq(pos, cond), stack1) = state.stack.pop(2)
    
    if(cond.isZero)
      state.withStack(stack1).step()
    else if (state.program.isValidJumpDest(pos.intValue))
      state.withStack(stack1).goto(pos.intValue)
    else
      state.withError(InvalidJump)
```